### PR TITLE
libstd => 2018

### PR DESCRIPTION
**Pull request by taiki-e**
Transitions `libstd` to Rust 2018; cc #58099

r? @Centril

### CODE REVIEW

File: src/libstd/lib.rs
Comment (taiki-e) [inline src/libstd/lib.rs]: It is hard work to change all at once (and review it), so `elided_lifetimes_in_paths` is allowed.

I will work on it after this PR is merged.

**Comment by taiki-e** [comment]
I think this error is very similar to [the error](https://github.com/rust-lang/rust/pull/58110#issuecomment-460063880) that occurred in #58110.

File: src/libstd/lib.rs
Comment (Centril) [reply src/libstd/lib.rs]: That sounds like a good idea; I think we want to keep `#![allow(explicit_outlives_requirements)]` around for this crate.

**Review by Centril — APPROVED** [review]
Modulo the ICE, everything looks good.

**Comment by taiki-e** [comment]
`libstd` has many codes depending on each target. Since testing with PR is done only on some targets, there may be an oversight somewhere. I think the first test on the `auto` branch may fail (If errors are found, I will fix it immediately).

**Comment by petrochenkov** [comment]
This is an ICE that was fixed in https://github.com/rust-lang/rust/pull/58058 if I'm not mistaken.

**Comment by taiki-e** [comment]
> This is an ICE that was fixed in #58058 if I'm not mistaken.

Thanks for the info!

**Comment by Centril** [comment]
@taiki-e seems your rebase  went a bit wrong... :)

**Comment by taiki-e** [comment]
@Centril Would you please rebase #58110?

#58253 has similar ICE too , so I would like to gather information from other PRs that have similar ICE.
Currently I do not know if this is a problem unique to my PRs...

**Comment by Centril** [comment]
@taiki-e Will do now that I'm back from All Hands... :)

**Comment by Centril** [comment]
@taiki-e Try to rebase now and let's see what happens... :)

**Comment by taiki-e** [comment]
@Centril Thanks for fix ICE!

**Comment by taiki-e** [comment]
Currently, occurring errors are the following two places:

* [test/ui/hygiene/no_implicit_prelude](https://github.com/rust-lang/rust/blob/master/src/test/ui/hygiene/no_implicit_prelude.rs)
* [test/ui/tag-that-dare-not-speak-its-name](https://github.com/rust-lang/rust/blob/master/src/test/ui/tag-that-dare-not-speak-its-name.rs)

 Related to #55630 and #56389.

**Comment by taiki-e** [comment]
@petrochenkov Should these tests be rewritten to work with Rust 2018? Or should these tests be tested with Rust 2015 (probably add `edition:2015` to header commands?)?

**Comment by petrochenkov** [comment]
@taiki-e 
For `hygiene/no_implicit_prelude` this is an expected change, an `ERROR` annotation needs to be added.
`tag-that-dare-not-speak-its-name` needs to be "fixed to work", probably by using `std::panic`.

**Comment by taiki-e** [comment]
CI passed. Thanks for your help @petrochenkov!

**Comment by Centril** [comment]
LGTM! @bors r+ p=1

**Comment by kennytm** [comment]
This time the error is legit.

**Comment by taiki-e** [comment]
Probably https://github.com/rust-lang-nursery/stdsimd/commit/a9bcbf4f67f6f806001046dcacc2403469f59b95#diff-793a5a022740a3025a89997a23d95cad.

**Comment by Centril** [comment]
cc @gnzlbg ^

**Comment by taiki-e** [comment]
I wrote a PR to solve this problem: https://github.com/rust-lang-nursery/stdsimd/pull/693

**Comment by gnzlbg** [comment]
Wait, did something change? How come now we are getting different errors ?

**Comment by Centril** [comment]
@gnzlbg I'm not seeing it... what changed in the errors?

**Comment by gnzlbg** [comment]
I think I might have a different PR in mind. I just recall some documentation errors due to how the new ARM ACLE module is structured.

**Comment by gnzlbg** [comment]
@taiki-e I think the stdsimd PR landed, this might need to update stdsimd to the latest commit on the master branch, but maybe it is worth it to wait till we fix the documentation build issue for core::arch::arm/aarch64.

**Comment by taiki-e** [comment]
(force-pushed to make sure the rebase did not fail)

**Comment by gnzlbg** [comment]
This is blocked on https://github.com/rust-lang/rust/pull/58675

**Comment by Centril** [comment]
@taiki-e https://github.com/rust-lang/rust/pull/58675 landed, can you rebase? :)

**Comment by taiki-e** [comment]
Rebased.

**Pull request merged**
